### PR TITLE
ReferenceGenome parsing through JSON in compiled functions

### PR DESCRIPTION
**Pull request by catoverdrive**

### CODE REVIEW

File: src/main/scala/is/hail/variant/ReferenceGenome.scala [hunk 1]
```diff
@@ -443,9 +444,43 @@ case class ReferenceGenome(name: String, contigs: Array[String], lengths: Map[St
...
+    implicit val formats: Formats = defaultJSONFormats
+    Serialization.write(jrg)
+  }
+
+  def addAsField(fb: FunctionBuilder[_]): (ClassFieldRef[ReferenceGenome], Code[Unit]) = {
+    val json = toJSONString
+    val chunkSize = 65535
+    val nChunks = (json.length() - 1) / chunkSize + 1
+    assert(nChunks > 0)
+    val stringRef = fb.newField[String](s"strrg_$name")
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/ReferenceGenome.scala]: I'm not sure there's a guarantee that RGs have no spaces. Would that cause a problem here?

File: src/main/scala/is/hail/variant/ReferenceGenome.scala [hunk 2]
```diff
@@ -443,9 +444,43 @@ case class ReferenceGenome(name: String, contigs: Array[String], lengths: Map[St
...
+      contigs.map(contig => JSONExtractContig(contig, contigLength(contig))),
+      xContigs, yContigs, mtContigs,
+      par.map(i => JSONExtractIntervalLocus(i.start.asInstanceOf[Locus], i.end.asInstanceOf[Locus])))
+    implicit val formats: Formats = defaultJSONFormats
+    Serialization.write(jrg)
+  }
+
+  def addAsField(fb: FunctionBuilder[_]): (ClassFieldRef[ReferenceGenome], Code[Unit]) = {
+    val json = toJSONString
+    val chunkSize = 65535
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/ReferenceGenome.scala]: prefer `(1 << 16) - 1` for clarity

File: src/main/scala/is/hail/variant/ReferenceGenome.scala [hunk 3]
```diff
@@ -443,9 +444,43 @@ case class ReferenceGenome(name: String, contigs: Array[String], lengths: Map[St
...
+    val stringAssembler = if (chunks.length == 1) {
+      stringRef := chunks.head
+    } else {
+      Code(stringRef := chunks.head,
+        Code(chunks.tail.map { s =>
+            stringRef := stringRef.invoke[String, String]("concat", s)
+          }: _*))
+    }
+
+    val field = fb.newField[ReferenceGenome](s"rg_$name")
```
Comment (tpoterba) [inline src/main/scala/is/hail/variant/ReferenceGenome.scala]: same as above

**Comment by catoverdrive** [comment]
the `loader` code is currently just a chunk of code---it gets run whenever it's invoked. When I put it into emit, I was going to protect it from getting run every time the function was evaluated with some checking, but I can put it directly into this function instead--that might be better.

**Comment by catoverdrive** [comment]
ugh actually doing this is not super straightforward! so unless you have super strong objections to this I'll probably jut put the logic directly into emit.

**Comment by tpoterba** [comment]
does that mean the genome reference is getting parsed once per evaluation? That's a pretty expensive thing to do. I'm OK with going this route for now, but we should make a list of things that we need to fix asap when our infrastructure allows.

**Comment by tpoterba** [comment]
can we do something like populate some static object with the RG, and try to look it up?

**Comment by catoverdrive** [comment]
no, its just a question of where we put the checking logic. When we add a reference genome in Emit, we'll probably just have to set some boolean flag that tells us if something has already been set, and only call the `loader` code if it hasn't been set.

**Comment by tpoterba** [comment]
ah, ok. Fair enough.

**Pull request merged**